### PR TITLE
Improve robustness of our devtools Gradle code

### LINKED ISSUE #11303: Improve robustness of our devtools Gradle code
**Description**
I was rebasing #9353 with recent fix from @gastaldi #11253, when I just stopped because it was too much for me to take in, the problem is not with @gastaldi or @mgorniew contribution which are great, our Gradle Devtool code is just too messy in its root. You can find most of it in `AbstractGradleBuildFile`. This is really too fragile: add a space to a build.gradle and it doesn't work anymore.

There are 3 really bad practices in this files:
1. We do the checks from formatted String
2. We format the snippets with some huge StringBuilders
3. We deal with the `build.gradle` model as a String

I think we could easily improve 1. and 2., 3 maybe a more complex. For 1., we could use regexp and for 2. we could use some Qute snippets template.

I believe, we should wait for this to merge #9353.

### PR DESCRIPTION
**Pull request by mgorniew**
Attempt to improve dependencies handling for Gradle. This leverages QuarkusModel from Gradle plugin to extract list of dependencies for project.

Fixes: https://github.com/quarkusio/quarkus/issues/11303

### CODE REVIEW

File: independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/project/buildfile/AbstractGradleBuildFile.java [hunk 1]
```diff
@@ -57,23 +56,17 @@ public void writeToDisk() throws IOException {
...
         if (!managed &&
                 (coords.getVersion() != null && !coords.getVersion().isEmpty())) {
             newDependency.append(":").append(coords.getVersion());
         }
+        boolean isBOM = "pom".equals(coords.getType());
+        if (isBOM && !managed) {
+            return String.format("enforcedPlatform(%s)", newDependency.append(quoteChar).toString());
+        }
         return newDependency.append(quoteChar).toString();
     }
```
[the remark this answers is not in the document]
Comment (mgorniew) [reply independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/project/buildfile/AbstractGradleBuildFile.java:72]: Yes, I think we can remove this check now.

[the remark this answers is not in the document]
Comment (ia3andy) [reply independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/project/buildfile/AbstractGradleBuildFile.java:72]: @mgorniew ok I am waiting for this PR to be "ready for review" with this change, let me know ;-)

File: devtools/gradle/src/main/java/io/quarkus/gradle/ConnectorDependencyResolver.java [hunk 2]
```diff
@@ -3,38 +3,37 @@
...
-import org.gradle.tooling.model.eclipse.EclipseExternalDependency;
-import org.gradle.tooling.model.eclipse.EclipseProject;
+
+import io.quarkus.bootstrap.resolver.QuarkusGradleModelFactory;
+import io.quarkus.bootstrap.resolver.model.QuarkusModel;
+import io.quarkus.runtime.LaunchMode;
 
 public final class ConnectorDependencyResolver {
 
     private List<Dependency> dependencies = null;
```
Comment (ia3andy) [inline devtools/gradle/src/main/java/io/quarkus/gradle/ConnectorDependencyResolver.java:18]: @mgorniew why are we using `org.apache.maven.model.Dependency` here?

Comment (mgorniew) [reply devtools/gradle/src/main/java/io/quarkus/gradle/ConnectorDependencyResolver.java:18]: This is model used by `io.quarkus.devtools.project.buildfile.BuildFile`.

Comment (ia3andy) [reply devtools/gradle/src/main/java/io/quarkus/gradle/ConnectorDependencyResolver.java:18]: @mgorniew I think we can easily change that :) just remove `getDependencies` and have `getDependenciesKeys` abstract, and move the current `getDependenciesKeys` logic to the `MavenBuildFile`, then you can avoid this dependency in Gradle..

Comment (mgorniew) [reply devtools/gradle/src/main/java/io/quarkus/gradle/ConnectorDependencyResolver.java:18]: `AppArtifactKey` doesn't contain version and there is `getInstalled` method which uses `getDependencies` to create `AppArtifactCoords`.

Comment (ia3andy) [reply devtools/gradle/src/main/java/io/quarkus/gradle/ConnectorDependencyResolver.java:18]: Then use `AppArtifactCoords` isnted of keys :)

Comment (aloubyansky) [reply devtools/gradle/src/main/java/io/quarkus/gradle/ConnectorDependencyResolver.java:18]: The issue is that we are using Maven's model Dependency in QuarkusPlatformDescriptor. So either way, we are going to be converting unless we change it in QuarkusPlatformDescriptor and everywhere else.

Comment (aloubyansky) [reply devtools/gradle/src/main/java/io/quarkus/gradle/ConnectorDependencyResolver.java:18]: It could put a start to it though.

Comment (mgorniew) [reply devtools/gradle/src/main/java/io/quarkus/gradle/ConnectorDependencyResolver.java:18]: Sure, I will change maven `Dependency` to `AppArtifactCoords`.

File: devtools/gradle/src/main/java/io/quarkus/gradle/builder/QuarkusModelBuilder.java [hunk 3]
```diff
@@ -101,7 +104,9 @@ public Object buildAll(String modelName, ModelParameter parameter, Project proje
 
         return new QuarkusModelImpl(new WorkspaceImpl(appArtifactCoords, getWorkspace(project.getRootProject(), mode)),
                 new LinkedList<>(appDependencies.values()),
-                extensionDependencies);
+                extensionDependencies,
+                deploymentDeps.stream().map(QuarkusModelBuilder::toEnforcedPlatformDependency)
+                        .filter(Objects::nonNull).collect(Collectors.toList()));
```
[the remark this answers is not in the document]
Comment (ia3andy) [reply devtools/gradle/src/main/java/io/quarkus/gradle/builder/QuarkusModelBuilder.java:109]: @aloubyansky I didn't really get what you mean here, @mgorniew maybe you did?

[the remark this answers is not in the document]
Comment (aloubyansky) [reply devtools/gradle/src/main/java/io/quarkus/gradle/builder/QuarkusModelBuilder.java:109]: It looks ok now. Alternatively, @mgorniew getEnforcedPlatforms(project) at line 88 returns the enforced platforms. So at that point you could simply make a copy of that set.

File: integration-tests/devtools/src/test/java/io/quarkus/devtools/commands/AddGradleExtensionsTest.java [hunk 4]
```diff
@@ -71,18 +70,26 @@ public TestingGradleBuildFile(Path projectDirPath, QuarkusPlatformDescriptor pla
...
             while (matcher.find()) {
-                final Dependency dep = new Dependency();
-                dep.setGroupId(matcher.group(1));
-                dep.setArtifactId(matcher.group(2));
-                dep.setVersion(matcher.group(3));
-                builder.add(dep);
+                builder.add(createDependency(matcher.group(1), matcher.group(2), matcher.group(3), "jar"));
+            }
+            if (getBuildContent().contains(
+                    "implementation enforcedPlatform(\"${quarkusPlatformGroupId}:${quarkusPlatformArtifactId}:${quarkusPlatformVersion}\")")) {
```
Comment (gastaldi) [inline integration-tests/devtools/src/test/java/io/quarkus/devtools/commands/AddGradleExtensionsTest.java:81]: We need to make this less brittle in the future (maybe regexp?)

Comment (aloubyansky) [reply integration-tests/devtools/src/test/java/io/quarkus/devtools/commands/AddGradleExtensionsTest.java:81]: In the future it should be checking for any specific property names at all.

Comment (aloubyansky) [reply integration-tests/devtools/src/test/java/io/quarkus/devtools/commands/AddGradleExtensionsTest.java:81]: Ah, it's a test, sorry. Ignore my comment above.

Comment (ia3andy) [reply integration-tests/devtools/src/test/java/io/quarkus/devtools/commands/AddGradleExtensionsTest.java:81]: @gastaldi this is just for testing ;-)

**Review by ia3andy — CHANGES_REQUESTED** [review]

**Comment by ia3andy** [comment]
@mgorniew this PR is still draft, is it ready?

**Comment by ia3andy** [comment]
cc @gsmet @sebersole

**Comment by ia3andy** [comment]
@mgorniew are you still on that one?

**Comment by mgorniew** [comment]
Yes, sorry. I will look into it today.

wt., 15 wrz 2020, 08:26 użytkownik Andy Damevin <notifications@github.com>
napisał:

**Comment by ia3andy** [comment]
@mgorniew if not a draft anymore can you switch it "ready for review"?

**Review by ia3andy — APPROVED** [review]
LGTM

**Pull request merged**